fix: Fixed Comments on Bitbucket Data Center & Cloud

zakisk opened the pull request:
when PipelineRun finishes PAC creates final status as comment on Bitbucket Data Center and Cloud, but as Bitbucket (both) doesn't support HTML comment they were not being rendered as HTML rather they were only raw HTML string. in this PR markdown is used as formatting template to prettify comment on Bitbucket.

Fixes #1876 alongside 

https://issues.redhat.com/browse/SRVKP-7331

# Changes <!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

<!-- Describe your changes here- ideally you can get that description straight from your descriptive commit message(s)! -->

# Submitter Checklist

- [x] 📝 Ensure your commit message is clear and informative. Refer to the How to write a git commit message guide. Include the commit message in the PR body rather than linking to an external site (e.g., Jira ticket).

- [x] ♽ Run make test lint before submitting a PR to avoid unnecessary CI processing. Consider installing [pre-commit](https://pre-commit.com/) and running pre-commit install in the repository root for an efficient workflow.

- [x] ✨ We use linters to maintain clean and consistent code. Run make lint before submitting a PR. Some linters offer a --fix mode, executable with make fix-linters (ensure [markdownlint](https://github.com/DavidAnson/markdownlint) and [golangci-lint](https://github.com/golangci/golangci-lint) are installed).

- [ ] 📖 Document any user-facing features or changes in behavior.

- [ ] 🧪 While 100% coverage isn't required, we encourage unit tests for code changes where possible.

- [ ] 🎁 If feasible, add an end-to-end test. See [README](https://github.com/openshift-pipelines/pipelines-as-code/blob/main/test/README.md) for details.

- [ ] 🔎 Address any CI test flakiness before merging, or provide a valid reason to bypass it (e.g., token rate limitations).

- If adding a provider feature, fill in the following details:

| Git Provider          | Supported |
|-----------------------|-----------|
| GitHub App            | ❌️        |
| GitHub Webhook        | ❌️        |
| Gitea                 | ❌️        |
| GitLab                | ❌️        |
| Bitbucket Cloud       | ✅️        |
| Bitbucket Data Center | ✅️        |

  (update the documentation accordingly)

aThorp96 commented on pkg/pipelineascode/pipelineascode.go:
```diff
@@ -226,7 +227,11 @@ func (p *PacRun) startPR(ctx context.Context, match matcher.Match) (*tektonv1.Pi
+	if strings.Contains(pr.GetAnnotations()[keys.GitProvider], "bitbucket") {
+		template = formatting.StartingPipelineRunMarkdown
+	}
+	msg, err := mt.MakeTemplate(template)
```
Non blocking suggestion: we have some shared provider logic in [provider.go](https://github.com/openshift-pipelines/pipelines-as-code/blob/main/pkg/provider/provider.go) regarding comments. Would you be opposed to putting this "change template depending on the provider implementation" in `provider.go`? Ideally code at this level wouldn't have any provider-specific code

aThorp96 commented on pkg/reconciler/reconciler.go:
```diff
@@ -276,7 +276,12 @@ func (r *Reconciler) updatePipelineRunToInProgress(ctx context.Context, logger *
+	if strings.Contains(pr.GetAnnotations()[keys.GitProvider], "bitbucket") {
+		template = formatting.StartingPipelineRunMarkdown
+	}
+	msg, err := mt.MakeTemplate(template)
```
Here too I think we could reduce the necessary code and avoid leaking the provider abstraction by using a central, shared provider template fetcher. E.g. something like this:
```go
// pkg/provider/provider.go
type CommentType int

const (
    StartingPipelineType CommentType = iota
    PipelineRunStatusType
    QueueingPipelineType
)

const htmlCommentTemplates map[CommentType]string{
    StartingPipelineType: formatting.StartingPipelineRunHTML,
    PipelineRunStatusType: formatting.PipelineRunStatusHTML,
    QueingPipelineType: formatting.QueingPipelineRunHTML,
}
const markdownCommentTemplates map[CommentType]string{
    StartingPipelineType: formatting.StartingPipelineRunMarkdown,
    PipelineRunStatusType: formatting.PipelineRunStatusMarkdown,
    QueingPipelineType: formatting.QueingPipelineRunMarkdown,
}

const markdownRequiredProviders []string{"bitbucket"}

func GetCommentTemplate(gitProvider string, commentType CommentType) string {
    templates := htmlCommentTemplates
    if slices.Contains(markdownRequiredProviders, gitProvider) {
        templates = markdownCommentTemplates
    }
    return templates[commentType]
}
```

And then in `reconciler.go` or `pipelineascode.go` you don't need to implement knowledge of which provider supports what
```go
template := provider.GetCommentTemplate(pr.GetAnnotations()[keys.GitProvider], provider.StartingPipelineType)
msg, err := mt.MakeTemplate(template)
```

zakisk replied:
@aThorp96 after some consideration I made changes, PTAL.

zakisk replied:
FYI, slices, maps can't be constant in Golang

zakisk committed:
update: Set `remeber-ok-to-test` settings to false by default

`remeber-ok-to-test` setting is set to false by default to mitigate
security risks of enabling an attacker to gain trust of repo owner
and push malicious code.

fixes: #1876

https://issues.redhat.com/browse/SRVKP-7238

Signed-off-by: Zaki Shaikh <zashaikh@redhat.com>

chmouel commented:
I was actually expecting another implementation from this PR, where you change the provider interface wiht a method like GetTemplate(type) and the provider itself will output whatever we want, so we don't have to do those if provider === "bitbucket" thing in the provider.GetCommentTamplate or mixing provider specific in the main code...

Let me know what do you think

zakisk commented:
> I was actually expecting another implementation from this PR, where you change the provider interface wiht a method like GetTemplate(type) and the provider itself will output whatever we want, so we don't have to do those if provider === "bitbucket" thing in the provider.GetCommentTamplate or mixing provider specific in the main code...
> 
> Let me know what do you think

Ah, you're right it looks now more concise

aThorp96 commented on pkg/provider/bitbucketcloud/bitbucket.go:
```diff
@@ -297,3 +302,7 @@ func (v *Provider) GetFiles(_ context.Context, _ *info.Event) (changedfiles.Chan
 	return "", nil
 }
+
+func (v *Provider) GetTemplate(commentType provider.CommentType) string {
+	return provider.GetMarkdownTemplate(commentType)
+}
```
Mmm, I like this much better as well. It feels more pragmatic for Go.

aThorp96 reviewed (approved):
:shipit:

chmouel commented:
/merge
